### PR TITLE
Optimize mining checks for various stake transactions

### PR DESCRIPTION
**Pull request by cjepson**
Many of the checks for the type of stake transaction in the main sorting
loop for mempool transactions of NewBlockTemplate were redundant. Now the
cached data for the transaction type is used instead. An expensive lookup
for a transaction cache was also moved for efficiency.

### CODE REVIEW

File: mining.go [hunk 1]
```diff
@@ -1158,23 +1159,11 @@ mempoolLoop:
...
-			continue
-		}
-
 		// Need this for a check below for stake base input, and to check
 		// the ticket number.
-		isSSGen, _ := stake.IsSSGen(tx)
-
-		if isSSGen, _ := stake.IsSSGen(tx); isSSGen {
+		txType := txDesc.Type
+		isSSGen := txType == stake.TxTypeSSGen
```
Comment (jcvernaleo) [inline mining.go]: I know this is legal go, but I kind of think lines like that are very hard to read when going through it quickly.  Is there any clearer way to do it?

File: mining.go [hunk 2]
```diff
@@ -1344,13 +1345,13 @@ mempoolLoop:
 		tx := prioItem.tx
 
 		// Store if this is an SStx or not.
-		isSStx, err := stake.IsSStx(tx)
+		isSStx := prioItem.txType == stake.TxTypeSStx
```
Comment (jcvernaleo) [inline mining.go]: Same comment for these three as the one above.

**Comment by jcvernaleo** [comment]
Other than the comments above this looks fine.  Someone who mines needs to test for real though.

**Comment by cjepson** [comment]
I cleaned up the extra declaration for the first case, but I think it's still as clean as it gets in the second case. The alternative is declaring all the vars and then doing a switch on the txType, which I think is ugly, or just not using a bool for them altogether and using equates to constants, which I think is ugly too.

**Comment by alexlyp** [comment]
ntACK

**Comment by jcvernaleo** [comment]
@cjepson fair enough about the second case, it is choosing between ugly and ugly.  So just waiting for the tACK and good to go.

**Comment by jolan** [comment]
Mined 70 blocks with this, nothing out of the ordinary noticed.

tACK

**Pull request merged**